File: content/snippets/scala2/interview_lcci/17/10_find_majority_element.scala
object Lcci1710FindMajorityElement {
  def majorityElement(nums: Array[Int]): Int = {
    var candidate = 0
    var count = 0
    for (x <- nums) {
      if (count == 0) candidate = x
      count += (if (x == candidate) 1 else -1)
    }
    if (nums.count(_ == candidate) > nums.length / 2) candidate else -1
  }

  def main(args: Array[String]): Unit = {
    println(majorityElement(Array(1, 2, 5, 9, 5, 9, 5, 5, 5))) // 5
    println(majorityElement(Array(3, 2))) // -1
  }
}

File: content/snippets/cpp/interview_lcci/17/10_find_majority_element.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int candidate = 0, count = 0;
        for (int x : nums) {
            if (count == 0) candidate = x;
            count += (x == candidate) ? 1 : -1;
        }
        count = 0;
        for (int x : nums) {
            if (x == candidate) count++;
        }
        return count > (int)nums.size() / 2 ? candidate : -1;
    }
};

int main() {
    Solution sol;
    vector<int> nums1 = {1, 2, 5, 9, 5, 9, 5, 5, 5};
    cout << sol.majorityElement(nums1) << endl; // 5

    vector<int> nums2 = {3, 2};
    cout << sol.majorityElement(nums2) << endl; // -1
    return 0;
}
